design-document.md

```diff
@@ -232,6 +232,352 @@ Note: Two non-colluding aggregators --- one leader and one standard aggregator
+  select (PAVerifyStartReq.task.proto) { // Payload for each helper share
+    case prio: PrioVerifyStartReq payloads<0..2^16-1>;
+    case hits: HitsVerifyStartReq payloads<0..2^16-1>;
+  }
+} PAVerifyStartReq;
+```
+
+This message contains any number of encrypted helper shares: it might contain
+the shares uploaded by a single client, or it might contain the shares uploaded
+by multiple clients; we only require that all of the shares correspond to the
+same PA task and HPKE key configuration. This allows the aggregator to
+batch-process a number of inputs at once.
+
+Each helper share has a corresponding message payload contained in
+`PAVerifyStartReq.shares`. This includes any protocol-specific information the
+helper needs for input validation, e.g., the joint randomness used by the leader
+and helper for the protocol run.
+
+Note that a well-formed message contains as many payloads as shares, i.e., the
+sequence `PAVerifyStartReq.payloads` has the same number of elements as
+`PAVerifyStartReq.shares`. Moreover, the i-th payload should correspond to the
+i-th share for each i.
+
+**Response.**
+The helper handles HTTP requests to `[helper]/verify_start` as follows. It first
+checks for a well-formed `PAVerifyStartReq`. If not found, then the helper
+aborts and alerts the leader with "malformed verify start request".
+
+Next, it looks for PA parameters `PAParam` for which `PAVerifyStartReq.task ==
+PAParam.task`. If not found, then it aborts and alerts the leader with
+"unrecognized task".
+
+Next, it looks up the HPKE config and corresponding secret key associated with
+`PAVerifyStartReq.key_config_id`. If not found, then it aborts and alerts the
+leader with "unrecognized key config".
+
+Finally, for each pair of shares and message payloads, the helper does as
+follows. Let `share` denote the share. It computes the HPKE context as
+
+```
+context = SetupBaseR(share.enc, sk, PAParam.task)
+```
+
+where `sk` is the secret key corresponding to the HPKE key config. It then
+derives its proof share from `context` and `share.payload` as specified by the
+PA protocol. Finally, the helper uses the proof share and corresponding message
+to derive its response for the protocol run.
+
+The helper responds to the HTTP request from the leader with the following
+message:
+
+```
+struct {
+  PAProto proto;
+  bool resend_shares; // Whether the leader should resend shares
+  select (PAVerifyStartResp.proto) {
+    case prio: PrioVerifyStartResp payloads<0..2^16-1>;
+    case hits: HitsVerifyStartResp payloads<0..2^16-1>;
+  }
+} PAVerifyStartResp;
+```
+
+The i-th element of `PAVerifyStartResp.payloads` corresponds to the i-th element
+of `PAVerifyStartReq.payloads`. The boolean `resend_shares` is set to `true` if
+the helper wants the leader to retransmit its shares in the last step of the
+input-validation protocol. This allows the helper to implement the protocol
+statelessly, if desired.
```

Comment:
It also requires the leader to maintain a mapping between shares and messages it sent to the helper. That sort of state seems error prone to manage. Hmm... could this message indicate specifically which shares should be re-sent, if any?

Also, I wonder if we should even make this an option. State management is hard enough as-is, so perhaps requiring just one entity to do it is better than possibly allowing others to do it?

Reply:
One way or the other, I don't think `resend_shares` is an attribute of an individual `PAVerifyStartResp`. A helper server either has storage or not and so would presumably always send the same value of `resend_shares`. Should this be part of the `PATask` or `PAParam` that the leader and all helpers have to agree upon before the protocol starts?

Reply:
To clarify why this bit exists: the helper shares may be so big that the leader and helper agree to use state rather than have to retransmit them. If set, then the leader is supposed to retransmit *all* of the shares.

> It also requires the leader to maintain a mapping between shares and messages it sent to the helper. That sort of state seems error prone to manage. Hmm... could this message indicate specifically which shares should be re-sent, if any?

The intention is that the leader needs to re-send *all* of the shares if the bit is true, or *none* if the bit is false. I don't think we should bother with supporting anything in the middle.


> Also, I wonder if we should even make this an option. State management is hard enough as-is, so perhaps requiring just one entity to do it is better than possibly allowing others to do it?

Shares are likely to be fairly large, especially for heavy hitters. I think this is going to need to be an option.

Reply:
> One way or the other, I don't think `resend_shares` is an attribute of an individual `PAVerifyStartResp`. A helper server either has storage or not and so would presumably always send the same value of `resend_shares`. Should this be part of the `PATask` or `PAParam` that the leader and all helpers have to agree upon before the protocol starts?

That is an option, though the bit is more deployment-specific rather than protocol-specific, which is why I put it in the high level message.

Reply:
Quick follow-up: I'm fine with removing the `resend_shares` bit for simplicity and just require the leader to always resend. We can always add it later if it's needed.

Reply:
I think we should err on the side of making the protocol simpler, and hence remove the option. My intuition is that shares should be small for Prio (since there aren't actually shares but rather just a PRNG seed from which shares are derived).

Reply:
I plan to remove the bit.